File: backend/controllers/authController.js
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const fs = require('fs');
const path = require('path');
const db = require('../db');
const transporter = require('../utils/mailer');
const { JWT_SECRET } = require('../config/config');

exports.requestOtp = async (req, res) => {
    const { email } = req.body;

    if (!email) {
        return res.status(400).json({ success: false, message: 'Email is required.' });
    }

    const checkEmailSql = "SELECT * FROM users WHERE email = ?";
    db.query(checkEmailSql, [email], async (err, result) => {
        if (err) {
            console.error('Database error during email check:', err);
            return res.status(500).json({ success: false, message: 'An error occurred during email check.' });
        }
        if (result.length > 0) {
            return res.status(409).json({ success: false, message: 'Email already in use. Please try logging in.' });
        }

        const otp = Math.floor(100000 + Math.random() * 900000).toString();
        const expiresAt = new Date(Date.now() + (5 * 60 * 1000));

        const insertOtpSql = `
            INSERT INTO otps (email, otp_code, expires_at)
            VALUES (?, ?, ?)
            ON DUPLICATE KEY UPDATE
            otp_code = VALUES(otp_code), expires_at = VALUES(expires_at), created_at = CURRENT_TIMESTAMP
        `;
        db.query(insertOtpSql, [email, otp, expiresAt], async (dbErr) => {
            if (dbErr) {
                console.error('Error storing OTP in DB:', dbErr);
                return res.status(500).json({ success: false, message: 'An error occurred while storing OTP.' });
            }

            try {
                const emailTemplatePath = path.join(__dirname, '../../frontend', 'templates', 'otp_email.html');
                let emailHtml = fs.readFileSync(emailTemplatePath, 'utf8');
                emailHtml = emailHtml.replace('{{OTP_CODE}}', otp);

                await transporter.sendMail({
                    from: '"Leirad Noznag" <darielganzon2003@gmail.com>',
                    to: email,
                    subject: 'Your OTP for Registration',
                    html: emailHtml,
                    text: `Your One-Time Password (OTP) is: ${otp}. It is valid for 5 minutes. Do not share this with anyone.`,
                });
                res.json({ success: true, message: 'OTP sent successfully to ' + email });
            } catch (error) {
                console.error('Error sending email:', error);
                res.status(500).json({ success: false, message: 'Failed to send OTP. Mailer Error: ' + error.message });
            }
        });
    });
};

exports.requestPasswordResetOtp = async (req, res) => {
    const { email } = req.body;

    if (!email) {
        return res.status(400).json({ success: false, message: 'Email is required.' });
    }

    const checkUserSql = `SELECT id FROM users WHERE email = ?`;
    db.query(checkUserSql, [email], async (err, result) => {
        if (err) {
            console.error('Database error during user check for forgot password:', err);
            return res.status(500).json({ success: false, message: 'An error occurred.' });
        }
        if (result.length === 0) {
            return res.status(404).json({ success: false, message: 'Email not found.' });
        }

        const otp = Math.floor(100000 + Math.random() * 900000).toString();
        const expiresAt = new Date(Date.now() + (5 * 60 * 1000));

        const insertOtpSql = `
            INSERT INTO otps (email, otp_code, expires_at)
            VALUES (?, ?, ?)
            ON DUPLICATE KEY UPDATE
            otp_code = VALUES(otp_code), expires_at = VALUES(expires_at), created_at = CURRENT_TIMESTAMP
        `;
        db.query(insertOtpSql, [email, otp, expiresAt], async (dbErr) => {
            if (dbErr) {
                console.error('Error storing OTP for forgot password in DB:', dbErr);
                return res.status(500).json({ success: false, message: 'An error occurred while storing OTP.' });
            }

            try {
                const emailTemplatePath = path.join(__dirname, '../../frontend', 'templates', 'forgot_password_otp_email.html');
                let emailHtml = fs.readFileSync(emailTemplatePath, 'utf8');
                emailHtml = emailHtml.replace('{{OTP_CODE}}', otp);

                await transporter.sendMail({
                    from: '"Leirad Noznag" <darielganzon2003@gmail.com>',
                    to: email,
                    subject: 'Password Reset OTP',
                    html: emailHtml,
                    text: `Your One-Time Password (OTP) for password reset is: ${otp}. It is valid for 5 minutes. Do not share this with anyone.`,
                });
                res.json({ success: true, message: 'Password reset OTP sent successfully to ' + email });
            } catch (error) {
                console.error('Error sending password reset email:', error);
                res.status(500).json({ success: false, message: 'Failed to send password reset OTP. Mailer Error: ' + error.message });
            }
        });
    });
};

exports.verifyOtpAndRegister = (req, res) => {
    const { firstname, middlename, lastname, email, password, otp } = req.body;

    if (!firstname || !lastname || !email || !password || !otp) {
        return res.status(400).json({ success: false, message: 'All fields including OTP are required.' });
    }

    const getOtpSql = "SELECT otp_code, expires_at FROM otps WHERE email = ?";
    db.query(getOtpSql, [email], (err, otpResult) => {
        if (err) {
            console.error('Error retrieving OTP from DB:', err);
            return res.status(500).json({ success: false, message: 'An error occurred during OTP verification.' });
        }

        if (otpResult.length === 0) {
            return res.status(400).json({ success: false, message: 'Invalid or expired OTP.' });
        }

        const storedOtp = otpResult[0];
        const currentTime = new Date();

        if (storedOtp.otp_code !== otp || currentTime > storedOtp.expires_at) {
            const deleteOtpSql = "DELETE FROM otps WHERE email = ?";
            db.query(deleteOtpSql, [email], (deleteErr) => {
                if (deleteErr) console.error('Error deleting expired/invalid OTP:', deleteErr);
            });
            return res.status(400).json({ success: false, message: 'Invalid or expired OTP.' });
        }

        const checkEmailSql = "SELECT * FROM users WHERE email = ?";
        db.query(checkEmailSql, [email], (err, result) => {
            if (err) {
                console.error(err);
                res.status(500).json({ success: false, message: 'An error occurred.' });
                return;
            }
            if (result.length > 0) {
                res.status(409).json({ success: false, message: 'Email already in use.' });
                return;
            }

            bcrypt.hash(password, 10, (hashErr, hashedPassword) => {
                if (hashErr) {
                    console.error(hashErr);
                    res.status(500).json({ success: false, message: 'An error occurred during password hashing.' });
                    return;
                }

                const insertSql = "INSERT INTO users (firstname, middlename, lastname, email, password, profilePicture) VALUES (?, ?, ?, ?, ?, ?)";
                db.query(insertSql, [firstname, middlename, lastname, email, hashedPassword, 'images/default-profile.png'], (err, result) => {
                    if (err) {
                        console.error(err);
                        res.status(500).json({ success: false, message: 'An error occurred.' });
                        return;
                    }
                    const deleteOtpSql = "DELETE FROM otps WHERE email = ?";
                    db.query(deleteOtpSql, [email], (deleteErr) => {
                        if (deleteErr) console.error('Error deleting OTP after successful registration:', deleteErr);
                    });
                    const accessToken = jwt.sign({ id: result.insertId, email: email }, JWT_SECRET, { expiresIn: '1h' });
                    res.json({ success: true, message: 'Registration successful!', token: accessToken });
                });
            });
        });
    });
};

exports.verifyPasswordResetOtp = (req, res) => {
    const { email, otp } = req.body;

    if (!email || !otp) {
        return res.status(400).json({ success: false, message: 'Email and OTP are required.' });
    }

    const getOtpSql = `SELECT otp_code, expires_at FROM otps WHERE email = ?`;
    db.query(getOtpSql, [email], (err, otpResult) => {
        if (err) {
            console.error('Error retrieving OTP for password reset from DB:', err);
            return res.status(500).json({ success: false, message: 'An error occurred during OTP verification.' });
        }

        if (otpResult.length === 0) {
            return res.status(400).json({ success: false, message: 'Invalid or expired OTP.' });
        }

        const storedOtp = otpResult[0];
        const currentTime = new Date();

        if (storedOtp.otp_code !== otp || currentTime > storedOtp.expires_at) {
            return res.status(400).json({ success: false, message: 'Invalid or expired OTP.' });
        }

        const deleteOtpSql = "DELETE FROM otps WHERE email = ?";
        db.query(deleteOtpSql, [email], (deleteErr) => {
            if (deleteErr) console.error('Error deleting OTP after successful verification:', deleteErr);
        });

        res.json({ success: true, message: 'OTP verified successfully. You can now reset your password.' });
    });
};

exports.resetPassword = (req, res) => {
    const { email, newPassword, confirmNewPassword } = req.body;

    if (!email || !newPassword || !confirmNewPassword) {
        return res.status(400).json({ success: false, message: 'All fields are required.' });
    }

    if (newPassword !== confirmNewPassword) {
        return res.status(400).json({ success: false, message: 'New password and confirm password do not match.' });
    }

    bcrypt.hash(newPassword, 10, (hashErr, hashedPassword) => {
        if (hashErr) {
            console.error(hashErr);
            res.status(500).json({ success: false, message: 'An error occurred during password hashing.' });
            return;
        }

        const updatePasswordSql = "UPDATE users SET password = ? WHERE email = ?";
        db.query(updatePasswordSql, [hashedPassword, email], (updateErr, updateResult) => {
            if (updateErr) {
                console.error('Error updating password:', updateErr);
                return res.status(500).json({ success: false, message: 'An error occurred while resetting password.' });
            }
            if (updateResult.affectedRows === 0) {
                return res.status(404).json({ success: false, message: 'User not found.' });
            }
            const clearTokenSql = "UPDATE users SET token = NULL WHERE email = ?";
            db.query(clearTokenSql, [email], (clearTokenErr) => {
                if (clearTokenErr) {
                    console.error('Error clearing token after password reset:', clearTokenErr);
                }
                res.json({ success: true, message: 'Password has been reset successfully! Please log in with your new password.' });
            });
        });
    });
};

exports.login = (req, res) => {
    const { email, password } = req.body;
    const sql = "SELECT * FROM users WHERE BINARY email = ?";
    db.query(sql, [email], (err, result) => {
        if (err) {
            console.error(err);
            return res.status(500).json({ success: false, message: 'An error occurred.' });
        }
        if (result.length > 0) {
            const user = result[0];
            bcrypt.compare(password, user.password, (compareErr, isMatch) => {
                if (compareErr) {
                    console.error(compareErr);
                    return res.status(500).json({ success: false, message: 'An error occurred during password comparison.' });
                }
                if (isMatch) {
                    const expiresIn = '1h';
                    const accessToken = jwt.sign({ id: user.id, email: user.email }, JWT_SECRET, { expiresIn });

                    const updateTokenSql = "UPDATE users SET token = ? WHERE id = ?";
                    db.query(updateTokenSql, [accessToken, user.id], (updateErr) => {
                        if (updateErr) {
                            console.error('Error storing token in DB:', updateErr);
                            return res.status(500).json({ success: false, message: 'An error occurred during login.' });
                        }
                        res.json({
                            success: true,
                            message: 'Login successful!',
                            token: accessToken
                        });
                    });
                } else {
                    res.status(401).json({ success: false, message: 'Invalid credentials!' });
                }
            });
        } else {
            res.status(401).json({ success: false, message: 'Invalid credentials!' });
        }
    });
};

exports.logout = (req, res) => {
    const userId = req.user.id;
    const sql = "UPDATE users SET token = NULL WHERE id = ?";
    db.query(sql, [userId], (err) => {
        if (err) {
            console.error('Error clearing token from DB on logout:', err);
            return res.status(500).json({ success: false, message: 'An error occurred during logout.' });
        }
        res.json({ success: true, message: 'Logout successful!' });
    });
};

File: backend/controllers/auth_controller.py
import hashlib
import uuid
import re
import logging
import jwt
from flask import jsonify, request, session
from supabase_client import supabase
from config import Config
from utils.mailer import send_otp_email, send_password_reset_email

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def hash_password(password):
    salt = uuid.uuid4().hex
    return hashlib.sha256((password + salt).encode()).hexdigest(), salt

def verify_password(hashed_password, salt, password):
    return hashed_password == hashlib.sha256((password + salt).encode()).hexdigest()

def validate_email(email):
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(pattern, email) is not None

def signup():
    try:
        data = request.get_json()
        email = data.get('email')
        password = data.get('password')
        confirm_password = data.get('confirmPassword')
        name = data.get('name')

        # Validation
        if not email or not password or not confirm_password or not name:
            return jsonify({'success': False, 'message': 'All fields are required'}), 400

        if not validate_email(email):
            return jsonify({'success': False, 'message': 'Invalid email format'}), 400

        if password != confirm_password:
            return jsonify({'success': False, 'message': 'Passwords do not match'}), 400

        if len(password) < 6:
            return jsonify({'success': False, 'message': 'Password must be at least 6 characters'}), 400

        # Check if Supabase client is initialized
        if supabase is None:
            logger.error("Supabase client is not initialized")
            return jsonify({'success': False, 'message': 'Database connection error'}), 500

        # Check if user already exists
        existing_user = supabase.table('users').select('*').eq('email', email).execute()
        if existing_user.data:
            return jsonify({'success': False, 'message': 'Email already registered'}), 400

        # Hash password
        hashed_password, salt = hash_password(password)

        # Create user
        user_data = {
            'email': email,
            'password_hash': hashed_password,
            'salt': salt,
            'name': name
        }
        
        result = supabase.table('users').insert(user_data).execute()
        
        if result.data:
            return jsonify({'success': True, 'message': 'Account created successfully'}), 201
        else:
            return jsonify({'success': False, 'message': 'Failed to create account'}), 500
            
    except Exception as e:
        logger.error(f'Signup error: {str(e)}')
        return jsonify({'success': False, 'message': 'An error occurred during signup'}), 500

def login():
    try:
        data = request.get_json()
        email = data.get('email')
        password = data.get('password')

        # Validation
        if not email or not password:
            return jsonify({'success': False, 'message': 'Email and password are required'}), 400

        # Check if Supabase client is initialized
        if supabase is None:
            logger.error("Supabase client is not initialized")
            return jsonify({'success': False, 'message': 'Database connection error'}), 500

        # Get user
        result = supabase.table('users').select('*').eq('email', email).execute()
        users = result.data
        
        if not users:
            return jsonify({'success': False, 'message': 'Invalid credentials'}), 401

        user = users[0]
        hashed_password = user['password_hash']
        salt = user['salt']

        # Verify password
        if not verify_password(hashed_password, salt, password):
            return jsonify({'success': False, 'message': 'Invalid credentials'}), 401

        # Generate JWT token
        token_payload = {
            'id': user['id'],
            'email': user['email']
        }
        
        token = jwt.encode(token_payload, Config.JWT_SECRET, algorithm='HS256')
        
        # Store user info in session
        session['user_id'] = user['id']
        session['user_email'] = user['email']
        session['user_name'] = user['name']

        return jsonify({
            'success': True, 
            'message': 'Login successful',
            'token': token,
            'user': {
                'id': user['id'],
                'email': user['email'],
                'name': user['name']
            }
        }), 200
        
    except Exception as e:
        logger.error(f'Login error: {str(e)}')
        return jsonify({'success': False, 'message': 'An error occurred during login'}), 500

def logout():
    try:
        session.clear()
        return jsonify({'success': True, 'message': 'Logged out successfully'}), 200
    except Exception as e:
        logger.error(f'Logout error: {str(e)}')
        return jsonify({'success': False, 'message': 'An error occurred during logout'}), 500

def forgot_password():
    try:
        data = request.get_json()
        email = data.get('email')

        if not email:
            return jsonify({'success': False, 'message': 'Email is required'}), 400

        # Check if Supabase client is initialized
        if supabase is None:
            logger.error("Supabase client is not initialized")
            return jsonify({'success': False, 'message': 'Database connection error'}), 500

        # Check if user exists
        result = supabase.table('users').select('*').eq('email', email).execute()
        users = result.data
        
        if not users:
            # We don't reveal if the email exists for security reasons
            return jsonify({'success': True, 'message': 'If the email exists, a password reset link has been sent'}), 200

        user = users[0]
        
        # Generate OTP
        otp = str(uuid.uuid4().int)[:6]
        otp_hash, salt = hash_password(otp)

        # Store OTP in database
        otp_data = {
            'user_id': user['id'],
            'otp_hash': otp_hash,
            'salt': salt
        }
        
        # Delete any existing OTPs for this user
        supabase.table('password_resets').delete().eq('user_id', user['id']).execute()
        
        # Insert new OTP
        supabase.table('password_resets').insert(otp_data).execute()

        # Send OTP email
        email_sent = send_password_reset_email(email, user['name'], otp)
        
        if email_sent:
            return jsonify({'success': True, 'message': 'If the email exists, a password reset link has been sent'}), 200
        else:
            return jsonify({'success': False, 'message': 'Failed to send email'}), 500
            
    except Exception as e:
        logger.error(f'Forgot password error: {str(e)}')
        return jsonify({'success': False, 'message': 'An error occurred'}), 500

def reset_password():
    try:
        data = request.get_json()
        email = data.get('email')
        otp = data.get('otp')
        new_password = data.get('newPassword')
        confirm_password = data.get('confirmPassword')

        # Validation
        if not email or not otp or not new_password or not confirm_password:
            return jsonify({'success': False, 'message': 'All fields are required'}), 400

        if new_password != confirm_password:
            return jsonify({'success': False, 'message': 'Passwords do not match'}), 400

        if len(new_password) < 6:
            return jsonify({'success': False, 'message': 'Password must be at least 6 characters'}), 400

        # Check if Supabase client is initialized
        if supabase is None:
            logger.error("Supabase client is not initialized")
            return jsonify({'success': False, 'message': 'Database connection error'}), 500

        # Get user
        user_result = supabase.table('users').select('*').eq('email', email).execute()
        users = user_result.data
        
        if not users:
            return jsonify({'success': False, 'message': 'Invalid request'}), 400

        user = users[0]

        # Get OTP record
        otp_result = supabase.table('password_resets').select('*').eq('user_id', user['id']).execute()
        otp_records = otp_result.data
        
        if not otp_records:
            return jsonify({'success': False, 'message': 'Invalid or expired OTP'}), 400

        otp_record = otp_records[0]
        otp_hash = otp_record['otp_hash']
        salt = otp_record['salt']

        # Verify OTP
        if not verify_password(otp_hash, salt, otp):
            return jsonify({'success': False, 'message': 'Invalid or expired OTP'}), 400

        # Hash new password
        new_hashed_password, new_salt = hash_password(new_password)

        # Update password
        supabase.table('users').update({
            'password_hash': new_hashed_password,
            'salt': new_salt
        }).eq('id', user['id']).execute()

        # Delete OTP record
        supabase.table('password_resets').delete().eq('user_id', user['id']).execute()

        return jsonify({'success': True, 'message': 'Password reset successfully'}), 200
        
    except Exception as e:
        logger.error(f'Reset password error: {str(e)}')
        return jsonify({'success': False, 'message': 'An error occurred'}), 500